Firestore/Source/Local/FSTLRUGarbageCollector.mm

```diff
@@ -66,28 +80,91 @@ size_t size() const {
+    return "Garbage Collection skipped";
+  } else {
+    std::string desc = "LRU Garbage Collection:\n";
+    absl::StrAppend(&desc, "\tCounted targets in ", targetCountDurationMs, "ms\n");
+    absl::StrAppend(&desc, "\tDetermined least recently used ", sequenceNumbersCollected,
+                    " sequence numbers in ", upperBoundDurationMs, "ms\n");
+    absl::StrAppend(&desc, "\tRemoved ", targetsRemoved, " targets in ", removedTargetsDurationMs,
+                    "ms\n");
+    absl::StrAppend(&desc, "\tRemoved ", documentsRemoved, " documents in ",
+                    removedDocumentsDurationMs, "ms\n");
+    absl::StrAppend(&desc, "\tCompacted leveldb database in ", dbCompactionDurationMs, "ms\n");
+    absl::StrAppend(&desc, "Total duration: ", total_duration(), "ms");
+    return desc;
+  }
+}
 
 @implementation FSTLRUGarbageCollector {
   id<FSTLRUDelegate> _delegate;
+  FSTLruGcParams _params;
+  long _startTime;
+  long _lastRunTime;
 }
 
-- (instancetype)initWithQueryCache:(id<FSTQueryCache>)queryCache
-                          delegate:(id<FSTLRUDelegate>)delegate {
+- (instancetype)initWithDelegate:(id<FSTLRUDelegate>)delegate params:(FSTLruGcParams)params {
   self = [super init];
   if (self) {
-    _queryCache = queryCache;
     _delegate = delegate;
+    _params = std::move(params);
+    _lastRunTime = -1;
   }
   return self;
 }
 
+- (FSTLruGcResults)tryRunGcWithLiveTargets:(NSDictionary<NSNumber *, FSTQueryData *> *)liveTargets {
+  if (_params.minBytesThreshold == kFIRFirestorePersistenceCacheSizeUnlimited) {
+    return FSTLruGcResults::DidNotRun();
+  }
+
+  size_t currentSize = [self byteSize];
+  if (currentSize < _params.minBytesThreshold) {
+    // Not enough on disk to warrant collection. Wait another timeout cycle.
+    return FSTLruGcResults::DidNotRun();
+  } else {
+    return [self runGCWithLiveTargets:liveTargets];
+  }
+}
+
+- (FSTLruGcResults)runGCWithLiveTargets:(NSDictionary<NSNumber *, FSTQueryData *> *)liveTargets {
+  Timestamp start = Timestamp::Now();
+  int sequenceNumbers = [self queryCountForPercentile:_params.percentileToCollect];
+  // Cap at the configured max
+  if (sequenceNumbers > _params.maximumSequenceNumbersToCollect) {
+    sequenceNumbers = _params.maximumSequenceNumbersToCollect;
+  }
+  Timestamp countedTargets = Timestamp::Now();
+
+  ListenSequenceNumber upperBound = [self sequenceNumberForQueryCount:sequenceNumbers];
+  Timestamp foundUpperBound = Timestamp::Now();
+
+  int numTargetsRemoved =
+      [self removeQueriesUpThroughSequenceNumber:upperBound liveQueries:liveTargets];
+  Timestamp removedTargets = Timestamp::Now();
+
+  int numDocumentsRemoved = [self removeOrphanedDocumentsThroughSequenceNumber:upperBound];
+  Timestamp removedDocuments = Timestamp::Now();
+
+  [_delegate runPostCompaction];
+  Timestamp compactedDb = Timestamp::Now();
+
+  return FSTLruGcResults{
+      .didRun = YES,
+      .targetCountDurationMs = millisecondsBetween(start, countedTargets),
```

Comment:
Does anything besides logging really need these timing results? If not, I'd suggest just doing the logging here and leaving these out of the results structure.

Reply:
Done.